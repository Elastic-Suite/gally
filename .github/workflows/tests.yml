name: Tests

on:
  workflow_call:
    inputs:
      version:
        description: "Version of Gally's packages (ex: master or 1.0.x)"
        type: string
        default: 'master'
        required: false
      last_published_version:
        description: "Last release of the current version"
        required: false
        default: "dev-master"
        type: string
    secrets:
      COMPOSER_AUTH:
        required: true

jobs:
  tests:
    if: github.event.pull_request.draft == false
    name: Tests
    runs-on: ubuntu-latest
    
    env:
      back_branch: ${{ inputs.version }}
      front_branch: ${{ inputs.version == 'master' && 'main' || inputs.version }}
      composer_version: ${{ inputs.version == 'master' && 'dev-master' || (contains(inputs.version, 'x') && format('{0}-dev', inputs.version) || inputs.version) }}
    
    steps:
      - name: Checkout
        uses: actions/checkout@v2
        with:
          ref: ${{ env.back_branch }}
      - name: Checkout Gally admin
        uses: actions/checkout@v3
        with:
          repository: 'Elastic-Suite/gally-admin'
          ref: ${{env.front_branch}}
          path: front/gally-admin
      - name: update gally standard package
        working-directory: api
        run: composer require "gally/gally-standard:${{ env.composer_version }} as ${{ inputs.last_published_version }}"
      - name: Add premium package repository
        uses: php-actions/composer@v6
        with:
          php_version: "8.1"
          args: --working-dir api
          command: config repositories.gally-premium composer https://packagist.smile.fr/gally
        env:
          COMPOSER_AUTH: ${{ secrets.COMPOSER_AUTH }}
      - name: Install premium package
        uses: php-actions/composer@v6
        with:
          php_version: "8.1"
          args: --no-cache --working-dir api
          command: require gally/gally-premium:${{ env.composer_version }}
        env:
          COMPOSER_AUTH: ${{ secrets.COMPOSER_AUTH }}
      - name: Pull images
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml pull --ignore-pull-failures || true
      - name: Start services
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml up --build -d --wait
      - name: Check HTTP reachability
        id: http_reachability
        run: curl -v -o /dev/null http://localhost
      - name: HTTP Failure
        if: ${{ failure() && steps.tests.http_reachability == 'failure' }}
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml logs
      - name: Check HTTPS reachability
        id: https_reachability
        run: curl  -vk -o /dev/null https://localhost
      - name: HTTPS Failure
        if: ${{ failure() && steps.tests.https_reachability == 'failure' }}
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml logs
      - name: Create test database
        run: |
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console -e test doctrine:database:create
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console -e test doctrine:migrations:migrate --no-interaction
      - name: Doctrine Schema Validator
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console doctrine:schema:validate --skip-sync
      - name: Fixtures load
        run: |
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console gally:index:clear --no-interaction
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console hautelook:fixtures:load --no-interaction -vv
          docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=off -T php bin/console doctrine:fixtures:load --append --no-interaction
      - name: Coverage
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -e XDEBUG_MODE=coverage -T php bin/phpunit --coverage-clover=coverage/coverage.xml vendor/gally

      - name: Cleanup coverage file
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -T php sed -i 's:<file name="/srv/:<file name="/:' coverage/coverage.xml

      - if: ${{ github.event_name == 'pull_request' }}
        name: Download artifact
        uses: dawidd6/action-download-artifact@v2.14.1
        continue-on-error: true
        id: base_coverage_report
        with:
          workflow: .github/workflows/ci.yml
          branch: ${{ github.base_ref }}
          name: coverage-report
          path: coverage/base # Retrieve previous coverage file at the root directory, not inside api/coverage

      - name: Backend Coverage Report as Comment (Compared to a base)
        if: ${{ github.event_name == 'pull_request' && steps.base_coverage_report.outcome == 'success' }}
        uses: lucassabreu/comment-coverage-clover@main
        continue-on-error: true
        with:
          with-chart: false
          file: api/coverage/coverage.xml
          base-file: coverage/base/coverage.xml
          max-line-coverage-decrease: 10.0
          max-method-coverage-decrease: 10.0
          min-line-coverage: 75
          min-method-coverage: 75
          signature: "Backend Coverage compared to target branch"

      - name: Backend Coverage Report as Comment (Clean)
        uses: lucassabreu/comment-coverage-clover@v0.3.0
        if: ${{ (github.event_name == 'pull_request' && steps.base_coverage_report.outcome != 'success') || github.event_name == 'push'}}
        with:
          with-chart: false
          file: api/coverage/coverage.xml
          signature: "Backend Coverage of this branch"

      - uses: actions/upload-artifact@v2
        with:
          name: coverage-report
          path: api/coverage/coverage.xml

      - name: Jest
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml exec -T pwa yarn test:ci

      - name: Frontend Coverage Report
        uses: 5monkeys/cobertura-action@v12
        if: ${{ github.event_name == 'pull_request' }}
        with:
          repo_token: "${{ secrets.GITHUB_TOKEN }}"
          report_name: "Frontend Coverage of this branch"
          path: front/example-app/coverage/cobertura-coverage.xml
          minimum_coverage: 1
          fail_below_threshold: false

      - name: "[End] Job failed, gathering logs"
        if: ${{ failure() }}
        run: docker compose -f docker-compose.yml -f docker-compose.ci.yml logs
