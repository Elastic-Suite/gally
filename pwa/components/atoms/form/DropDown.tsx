import { useRef, useState } from 'react'
import { SelectUnstyledProps } from '@mui/base/SelectUnstyled'
import { MultiSelectUnstyledProps } from '@mui/base/MultiSelectUnstyled'
import OptionUnstyled, { optionUnstyledClasses } from '@mui/base/OptionUnstyled'
import { styled } from '@mui/system'
import InputLabel from '@mui/material/InputLabel'
import FormControl from '@mui/material/FormControl'

import Checkbox from './Checkbox'
import MultiSelect from './MultiSelect'
import Select from './Select'

const Option = styled(OptionUnstyled)(({ theme }) => ({
  fontFamily: 'Inter',
  listStyle: 'none',
  padding: '8px 12px',
  cursor: 'default',
  fontWeight: 400,
  fontSize: 12,
  lineHeight: '18px',
  color: theme.palette.colors.neutral['800'],
  '&:last-of-type': {
    borderBottom: 'none',
  },
  [`&.${optionUnstyledClasses.selected}`]: {
    fontWeight: 'bold',
  },
  [`&.${optionUnstyledClasses.highlighted}`]: {
    backgroundColor: theme.palette.colors.neutral['200'],
  },
  [`&.${optionUnstyledClasses.disabled}`]: {
    color: theme.palette.colors.neutral['400'],
  },
  [`&:hover:not(.${optionUnstyledClasses.disabled})`]: {
    backgroundColor: theme.palette.colors.neutral['200'],
  },
}))

export interface IOption {
  disabled?: boolean
  id?: string
  label: string
  value: unknown
}

export type IOptions = IOption[]

interface ICommonProps {
  label?: string
  multiple?: boolean
  options: IOptions
  required?: boolean
}

export interface ISelectProps
  extends Omit<SelectUnstyledProps<unknown>, 'components'>,
    ICommonProps {
  multiple?: false
}

export interface IMultiSelectProps
  extends Omit<MultiSelectUnstyledProps<unknown>, 'components'>,
    ICommonProps {
  multiple: true
}

export type IDropDownProps = ISelectProps | IMultiSelectProps

export default function DropDown(props: IDropDownProps): JSX.Element {
  const { label, multiple, options, required, value, ...selectProps } = props
  const [listboxOpen, setListboxOpen] = useState(false)
  const ignoreClick = useRef(false)

  function handleCheckboxMouseDown(): void {
    if (listboxOpen) {
      ignoreClick.current = true
    }
  }

  function handleListOpenChange(state: boolean): void {
    if (!ignoreClick.current) {
      setListboxOpen(state)
    }
    ignoreClick.current = false
  }

  return (
    <FormControl variant="standard">
      {label ? (
        <InputLabel shrink required={required}>
          {label}
        </InputLabel>
      ) : null}
      {multiple ? (
        <MultiSelect
          listboxOpen={listboxOpen}
          onListboxOpenChange={handleListOpenChange}
          {...(selectProps as IMultiSelectProps)}
          value={value}
        >
          {options.map((option) => (
            <Option key={option.id || String(option.value)} {...option}>
              <Checkbox
                checked={value.includes(option.value)}
                label={option.label}
                list
                onMouseDown={handleCheckboxMouseDown}
              />
            </Option>
          ))}
        </MultiSelect>
      ) : (
        <Select {...(selectProps as ISelectProps)} value={value}>
          {!required && <Option value="">&nbsp;</Option>}
          {options.map((option) => (
            <Option key={option.id || String(option.value)} {...option}>
              {option.label}
            </Option>
          ))}
        </Select>
      )}
    </FormControl>
  )
}
