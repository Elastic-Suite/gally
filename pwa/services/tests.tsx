import { RenderOptions, render } from '@testing-library/react'
import { renderHook } from '@testing-library/react-hooks'
import { PreloadedState } from '@reduxjs/toolkit'
import { ReactChild } from 'react'

import AppProvider from '~/components/AppProvider'
import { AppStore, RootState, setupStore } from '~/store'

export interface IExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  preloadedState?: PreloadedState<RootState>
  store?: AppStore
}

export interface IRenderOutput {
  store: AppStore
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactChild }): JSX.Element {
    return <AppProvider store={store}>{children}</AppProvider>
  }
  return { store, ...render(ui, { wrapper: Wrapper, ...renderOptions }) }
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  {
    preloadedState = {},
    store = setupStore(preloadedState),
    ...renderOptions
  }: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactChild }): JSX.Element {
    return <AppProvider store={store}>{children}</AppProvider>
  }
  return { store, ...renderHook(hook, { wrapper: Wrapper, ...renderOptions }) }
}
