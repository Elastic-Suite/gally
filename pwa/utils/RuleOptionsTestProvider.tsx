import { ReactChild } from 'react'

import sourceFields from '~/public/mocks/source_fields.json'
import sourceFieldLabels from '~/public/mocks/source_field_labels.json'
import { RuleAttributeType } from '~/types'

import RuleOptionsProvider from '~/components/stateful-providers/RuleOptionsProvider/RuleOptionsProvider'

interface IProps {
  children: ReactChild
}

const sourceFieldLabelsMap = new Map(
  sourceFieldLabels['hydra:member'].map((label) => [
    label.sourceField,
    label.label,
  ])
)

const fields = sourceFields['hydra:member'].map((field) => {
  const label = sourceFieldLabelsMap.get(`/source_fields/${field.id}`)
  return {
    id: field.id,
    code: field.code,
    label: (label || field.defaultLabel) ?? '',
    type: field.type as RuleAttributeType,
  }
})

function RuleOptionsTestProvider(props: IProps): JSX.Element {
  const { children } = props
  return (
    <RuleOptionsProvider catalogId={-1} localizedCatalogId={-1} fields={fields}>
      {children}
    </RuleOptionsProvider>
  )
}

export default RuleOptionsTestProvider
