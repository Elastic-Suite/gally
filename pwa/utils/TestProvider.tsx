import { ReactChild } from 'react'

import { schemaContext, userContext } from '~/contexts'
import { IApi, IUser } from '~/types'

interface IProps {
  api: IApi
  children: ReactChild
}

const timestamp = Math.floor(Date.now() / 1000)
const expires = 8 * 60 * 60 // 8 hours

const user = {
  iat: timestamp,
  exp: timestamp + expires,
  roles: ['ROLE_ADMIN', 'ROLE_CONTRIBUTOR'],
  username: 'admin@example.com',
} as IUser

function TestProvider(props: IProps): JSX.Element {
  const { api, children } = props
  return (
    <userContext.Provider value={user}>
      <schemaContext.Provider value={api}>{children}</schemaContext.Provider>
    </userContext.Provider>
  )
}

export default TestProvider
