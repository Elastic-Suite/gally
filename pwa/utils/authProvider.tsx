import jwtDecode from "jwt-decode";
import { ENTRYPOINT } from "config/entrypoint";

export default {
  login: ({ username, password }) => {
    const request = new Request(
      `${ENTRYPOINT}/authentication_token`,
      {
        method: "POST",
        body: JSON.stringify({ email: username, password }),
        headers: new Headers({ "Content-Type": "application/json" }),
      }
    );
    return fetch(request)
      .then((response) => {
        if (response.status < 200 || response.status >= 300) {
          throw new Error(response.statusText);
        }
        return response.json();
      })
      .then(({ token }) => {
        localStorage.setItem("token", token);
        // @Todo: Ugly fix awaiting the real front of the application.
        // Allows to reload the page after the authentication, to avoid infinite loading message.
        location.reload();
      });
  },
  logout: () => {
    localStorage.removeItem("token");
    return Promise.resolve();
  },
  checkAuth: () => {
    try {
      // @ts-ignore
      if (!localStorage.getItem("token") || new Date().getTime() / 1000 > jwtDecode(localStorage.getItem("token"))?.exp) {
        return Promise.reject();
      }
      return Promise.resolve();
    } catch (e) {
      // override possible jwtDecode error
      return Promise.reject();
    }
  },
  checkError: (err) => {
    if ([401, 403].includes(err?.status || err?.response?.status)) {
      localStorage.removeItem("token");
      return Promise.reject();
    }
    return Promise.resolve();
  },
  getPermissions: () => Promise.resolve(),
};
