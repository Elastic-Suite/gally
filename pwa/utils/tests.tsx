import AppProvider from '~/components/AppProvider'

jest.mock('react-i18next', () => ({
  useTranslation: () => ({ t: (key) => key }),
}))

export const wrapper = ({ children }) => <AppProvider>{children}</AppProvider>
