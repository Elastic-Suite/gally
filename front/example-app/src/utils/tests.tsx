import { RenderOptions, render, renderHook } from '@testing-library/react'
import { MemoryRouter } from 'react-router-dom'
import { ReactNode } from 'react'
import { api } from 'gally-admin-shared'

import TestProvider from './TestProvider'

export type IExtendedRenderOptions = Omit<RenderOptions, 'queries'>

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderWithProviders(
  ui: React.ReactElement,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <MemoryRouter>
        <TestProvider api={api}>{children}</TestProvider>
      </MemoryRouter>
    )
  }
  return render(ui, { wrapper: Wrapper, ...renderOptions })
}

// eslint-disable-next-line @typescript-eslint/explicit-module-boundary-types, @typescript-eslint/explicit-function-return-type
export function renderHookWithProviders<R>(
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  hook: (props: any) => R,
  renderOptions: IExtendedRenderOptions = {}
) {
  function Wrapper({ children }: { children: ReactNode }): JSX.Element {
    return (
      <MemoryRouter>
        <TestProvider api={api}>{children}</TestProvider>
      </MemoryRouter>
    )
  }
  return renderHook(hook, { wrapper: Wrapper, ...renderOptions })
}
