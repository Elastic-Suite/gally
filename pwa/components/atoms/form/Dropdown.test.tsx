import { renderWithProviders } from '~/utils/tests'

import DropDown from './DropDown'

describe('DropDown match snapshot', () => {
  it('BadgeDisabledFalse', () => {
    const { container } = renderWithProviders(
      <DropDown
        disabled={false}
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Fourty', value: 40, disabled: true },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeDisabledTrue', () => {
    const { container } = renderWithProviders(
      <DropDown
        disabled
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Fourty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeLabel', () => {
    const { container } = renderWithProviders(
      <DropDown
        label="Label"
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Fourty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeRequiredFalse', () => {
    const { container } = renderWithProviders(
      <DropDown
        required={false}
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Fourty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeRequiredTrue', () => {
    const { container } = renderWithProviders(
      <DropDown
        required
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Fourty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })

  it('BadgeMultiple', () => {
    const { container } = renderWithProviders(
      <DropDown
        value={[]}
        multiple
        disabled={false}
        label="Label"
        options={[
          { label: 'Ten', value: 10 },
          { label: 'Twenty', value: 20 },
          { label: 'Thirty', value: 30 },
          { label: 'Fourty', value: 40 },
        ]}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
