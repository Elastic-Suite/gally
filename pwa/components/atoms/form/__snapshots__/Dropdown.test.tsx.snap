// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DropDown match snapshot BadgeDisabledFalse 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1c9gikl-MuiFormControl-root"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelectUnstyled-root css-k616if"
      type="button"
    >
      <span />
      <ion-icon
        name="chevron-down"
      />
    </button>
  </div>
</div>
`;

exports[`DropDown match snapshot BadgeDisabledTrue 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1c9gikl-MuiFormControl-root"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelectUnstyled-root Mui-disabled css-k616if"
      disabled=""
      type="button"
    >
      <span />
      <ion-icon
        name="chevron-down"
      />
    </button>
  </div>
</div>
`;

exports[`DropDown match snapshot BadgeLabel 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1c9gikl-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary css-ssr94h-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
    >
      Label
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelectUnstyled-root css-k616if"
      type="button"
    >
      <span />
      <ion-icon
        name="chevron-down"
      />
    </button>
  </div>
</div>
`;

exports[`DropDown match snapshot BadgeMultiple 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1c9gikl-MuiFormControl-root"
  >
    <label
      class="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-standard MuiFormLabel-colorPrimary css-ssr94h-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
    >
      Label
    </label>
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelectUnstyled-root css-k616if"
      type="button"
    >
      <span>
        
      </span>
      <ion-icon
        name="chevron-down"
      />
    </button>
  </div>
</div>
`;

exports[`DropDown match snapshot BadgeRequiredFalse 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1c9gikl-MuiFormControl-root"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelectUnstyled-root css-k616if"
      type="button"
    >
      <span />
      <ion-icon
        name="chevron-down"
      />
    </button>
  </div>
</div>
`;

exports[`DropDown match snapshot BadgeRequiredTrue 1`] = `
<div>
  <div
    class="MuiFormControl-root css-1c9gikl-MuiFormControl-root"
  >
    <button
      aria-expanded="false"
      aria-haspopup="listbox"
      class="MuiSelectUnstyled-root css-k616if"
      type="button"
    >
      <span />
      <ion-icon
        name="chevron-down"
      />
    </button>
  </div>
</div>
`;
